jiant/preprocess.py

```diff
@@ -552,20 +572,27 @@ def add_task_label_vocab(vocab, task):
         vocab.add_token_to_namespace(label, namespace)
 
 
-def add_bert_wpm_vocab(vocab, bert_model_name):
-    """Add BERT WPM vocabulary for use with pre-tokenized data.
+def add_pytorch_transformers_wpm_vocab(vocab, tokenizer_name):
+    """Add BERT/XLNet WPM vocabulary for use with pre-tokenized data.
 
-    BertTokenizer has a convert_tokens_to_ids method, but this doesn't do
-    anything special so we can just use the standard indexers.
+    These tokenizers have a convert_tokens_to_ids method, but this doesn't do
+    anything special, so we can just use the standard indexers.
```

Comment:
Do we know what this comment is about? I'm trying to figure out if there's anything we need to do differently now that we've moving from wordpiece to sentencepiece...

Reply:
For BERT at least, the implementation of `convert_tokens_to_ids` is pretty trivial:
```
def convert_by_vocab(vocab, items):
  """Converts a sequence of [tokens|ids] using the vocab."""
  output = []
  for item in items:
    output.append(vocab[item])
  return output


def convert_tokens_to_ids(vocab, tokens):
  return convert_by_vocab(vocab, tokens)
```
(from https://github.com/google-research/bert/blob/d66a146741588fb208450bde15aa7db143baaa69/tokenization.py#L144)

the indexers will produce the correct ids (up to an offset) as long as the input is properly /tokenized/ beforehand and everything is in the vocabulary (so the indexer will never return an AllenNLP OOV index).

Reply:
But there was some hacky code in `bert/utils.py` to correct for that offset (the ids - 2 line):
```
        # "Correct" ids to account for different indexing between BERT and	
        # AllenNLP.	
        # The AllenNLP indexer adds a '@@UNKNOWN@@' token to the	
        # beginning of the vocabulary, *and* treats that as index 1 (index 0 is	
        # reserved for padding).	
        ids[ids == 0] = self._pad_id + 2  # Shift the indices that were at 0 to become 2.	
        # Index 1 should never be used since the BERT WPM uses its own	
        # unk token, and handles this at the string level before indexing.	
        assert (ids > 1).all()	
        ids -= 2  # shift indices to match BERT wordpiece embeddings
```

so if we use the `tokenizer.convert_ids_to_tokens` and ignore indexers, we don't need to worry about that anymore. Worth setting a breakpoint inside the embedding module to sanity-check this, though.

Reply:
If we wanted to support GPT-2, XLNet, etc., does it make more sense to supporting their tokenizer interface (either by matching APIs or writing wrappers, I guess) ?

Reply:
Thanks! That code lives on—it's not in pytorch_transformers_interface. XLNet uses pretty much the same logic (tweaked slightly to support unknown chars), and I see no reason why GPT-2 would be all that different. I'll put that off for now unless anyone sees a reason that something like this would break on BERT/XLNet.